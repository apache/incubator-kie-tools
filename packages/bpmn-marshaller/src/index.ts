/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { Meta, getInstanceNs, domParser, getParser } from "@kie-tools/xml-parser-ts";
import {
  meta as bpmn20meta,
  elements as bpmn20elements,
  subs as bpmn20subs,
  root as bpmn20root,
  ns as bpmn20ns,
} from "./schemas/bpmn-2_0/ts-gen/meta";

import { BPMN20__tDefinitions } from "./schemas/bpmn-2_0/ts-gen/types";

type BpmnDefinitions = { definitions: BPMN20__tDefinitions };

type BpmnMarshaller = {
  parser: { parse(): BpmnDefinitions };
  builder: { build(json: BpmnDefinitions): string };
  instanceNs: Map<string, string>;
  root: { element: string; type: string };
  meta: Meta;
  version: "2.0";
};

export function getMarshaller(xml: string): BpmnMarshaller {
  const domdoc = domParser.getDomDocument(xml);
  const instanceNs = getInstanceNs(domdoc);

  const p = getParser<BpmnDefinitions>({
    ns: bpmn20ns,
    meta: bpmn20meta,
    subs: bpmn20subs,
    elements: bpmn20elements,
    root: bpmn20root,
  });

  return {
    instanceNs,
    version: "2.0",
    root: bpmn20root,
    meta: bpmn20meta,
    parser: { parse: () => p.parse({ type: "domdoc", domdoc, instanceNs }).json },
    builder: { build: (json: BpmnDefinitions) => p.build({ json, instanceNs }) },
  };
}
