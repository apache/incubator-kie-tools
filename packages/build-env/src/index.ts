/*
 * Copyright 2022 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export type EnvAndVarsWithName<T> = { vars: { [K in keyof T]: VarWithName }; env: object };

export type Var = {
  default: string | undefined;
  description: string;
};

export type VarWithName = Var & { name: string };

//

export function str2bool(str: string | undefined) {
  return str === "true";
}

export function get(envVar: VarWithName) {
  return process.env[envVar.name];
}

export function getOrDefault(envVar: VarWithName) {
  return get(envVar) ?? envVar.default;
}

export function varsWithName<T>(obj: { [K in keyof T]: Var }) {
  [...Object.keys(obj)].forEach((key) => {
    (obj[key as keyof T] as VarWithName)["name"] = key;
  });

  return obj as { [K in keyof T]: VarWithName };
}

export function composeEnv<T>(deps: EnvAndVarsWithName<any>[], obj: EnvAndVarsWithName<T>) {
  return {
    vars: { ...obj.vars, ...deps.reduce((acc, d) => ({ ...acc, ...d.vars }), {}) },
    env: { ...obj.env, ...deps.reduce((acc, d) => ({ ...acc, ...d.env }), {}) },
  };
}
