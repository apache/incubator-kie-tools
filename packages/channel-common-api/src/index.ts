export * from "./ChannelType";
export * from "./ResourceContentOptions";
export * from "./ResourceListOptions";
export * from "./SearchType";
export * from "./ResourceContentRequest";
export * from "./OperatingSystem";
export * from "./ResourceContentService";
export * from "./KogitoEdit";
export * from "./ResourcesList";
export * from "./ResourceListRequest";
export * from "./ResourceContent";
export * from "./ContentType";
export * from "./KogitoCommonChannelApi";
