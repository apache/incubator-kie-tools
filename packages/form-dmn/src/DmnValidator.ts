/*
 * Copyright 2022 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as metaSchemaDraft04 from "ajv/lib/refs/json-schema-draft-04.json";
import { Validator } from "@kie-tools/form";
import { DmnFormI18n } from "./i18n";
import { DmnFormJsonSchemaBridge } from "./uniforms";

export const DAYS_AND_TIME =
  /^(-|\+)?P(?:([-+]?[0-9]*)D)?(?:T(?:([-+]?[0-9]*)H)?(?:([-+]?[0-9]*)M)?(?:([-+]?[0-9]*)S)?)?$/;
export const YEARS_AND_MONTHS = /^(-|\+)?P(?:([-+]?[0-9]*)Y)?(?:([-+]?[0-9]*)M)?$/;

export class DmnValidator extends Validator {
  private readonly SCHEMA_DRAFT4 = "http://json-schema.org/draft-04/schema#";

  constructor(i18n: DmnFormI18n) {
    super(i18n);
    this.setupValidator();
    this.i18n = i18n;
  }

  // Add meta schema v4 and period format
  private setupValidator() {
    this.ajv.addMetaSchema(metaSchemaDraft04);
    this.ajv.addFormat("days and time duration", {
      type: "string",
      validate: (data: string) => !!data.match(DAYS_AND_TIME),
    });

    this.ajv.addFormat("years and months duration", {
      type: "string",
      validate: (data: string) => !!data.match(YEARS_AND_MONTHS),
    });
  }

  // Override to add period validation
  public createValidator(jsonSchema: any) {
    const validator = this.ajv.compile(jsonSchema);

    return (model: any) => {
      // AJV doesn't handle dates objects. This transformation converts Dates to their UTC format.
      validator(JSON.parse(JSON.stringify(model)));

      if (!validator.errors?.length) {
        return null;
      }

      return {
        details: validator.errors?.map((error: any) => {
          if (error.keyword === "format") {
            if ((error.params as any).format === "days and time duration") {
              return { ...error, message: (this.i18n as DmnFormI18n).validation.daysAndTimeError };
            }
            if ((error.params as any).format === "years and months duration") {
              return { ...error, message: (this.i18n as DmnFormI18n).validation.yearsAndMonthsError };
            }
          }
          return error;
        }),
      };
    };
  }

  public getBridge(formSchema: object): DmnFormJsonSchemaBridge {
    const formDraft4 = { ...formSchema, $schema: this.SCHEMA_DRAFT4 };
    const validator = this.createValidator(formDraft4);
    return new DmnFormJsonSchemaBridge(formDraft4, validator, this.i18n as DmnFormI18n);
  }
}
