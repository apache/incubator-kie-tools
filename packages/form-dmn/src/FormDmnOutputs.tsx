/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import * as React from "react";
import { useCallback, useEffect, useMemo, useRef, useState } from "react";
import { CheckCircleIcon } from "@patternfly/react-icons/dist/js/icons/check-circle-icon";
import { InfoCircleIcon } from "@patternfly/react-icons/dist/js/icons/info-circle-icon";
import { ExclamationCircleIcon } from "@patternfly/react-icons/dist/js/icons/exclamation-circle-icon";
import { Tooltip } from "@patternfly/react-core/dist/js/components/Tooltip";
import { OutlinedQuestionCircleIcon } from "@patternfly/react-icons/dist/js/icons/outlined-question-circle-icon";
import {
  DescriptionList,
  DescriptionListDescription,
  DescriptionListGroup,
  DescriptionListTerm,
} from "@patternfly/react-core/dist/js/components/DescriptionList";
import { Card, CardBody, CardFooter, CardTitle } from "@patternfly/react-core/dist/js/components/Card";
import { Title } from "@patternfly/react-core/dist/js/components/Title";
import { EmptyState, EmptyStateBody, EmptyStateIcon } from "@patternfly/react-core/dist/js/components/EmptyState";
import { Text, TextContent } from "@patternfly/react-core/dist/js/components/Text";
import { formDmnI18n } from "./i18n";
import { I18nWrapped } from "@kie-tools-core/i18n/dist/react-components";
import "./styles.scss";
import { ErrorBoundary } from "@kie-tools/dmn-runner/dist/ErrorBoundary";
import { ExclamationTriangleIcon } from "@patternfly/react-icons/dist/js/icons/exclamation-triangle-icon";
import { DecisionResult, DmnEvaluationStatus, DmnEvaluationResult } from "@kie-tools/extended-services-api";

const ISSUES_URL = "https://github.com/apache/incubator-kie-issues/issues";

const DATE_REGEX = /\d{4}-(?:0[1-9]|1[0-2])-(?:0[1-9]|[1-2]\d|3[0-1])T(?:[0-1]\d|2[0-3]):[0-5]\d:[0-5]\dZ/;

enum FormDmnOutputsStatus {
  EMPTY,
  ERROR,
  VALID,
}

type DeepPartial<T> = {
  [P in keyof T]?: DeepPartial<T[P]>;
};

export interface FormDmnOutputsProps {
  results?: DecisionResult[];
  differences?: Array<DeepPartial<DecisionResult>>;
  locale?: string;
  notificationsPanel: boolean;
  openExecutionTab?: () => void;
}

export function FormDmnOutputs({ openExecutionTab, ...props }: FormDmnOutputsProps) {
  const [formResultStatus, setFormResultStatus] = useState<FormDmnOutputsStatus>(FormDmnOutputsStatus.EMPTY);
  const [formResultError, setFormResultError] = useState<boolean>(false);
  const i18n = useMemo(() => {
    formDmnI18n.setLocale(props.locale ?? navigator.language);
    return formDmnI18n.getCurrent();
  }, [props.locale]);
  const errorBoundaryRef = useRef<ErrorBoundary>(null);

  useEffect(() => {
    props.differences?.forEach((difference, index) => {
      if (Object.keys(difference).length === 0) {
        return;
      }

      const updatedResult = document.getElementById(`${index}-dmn-result`);
      updatedResult?.classList.add("kogito--editor__dmn-form-result__leaf-updated");
    });
  }, [props.differences]);

  const onAnimationEnd = useCallback((e: React.AnimationEvent<HTMLElement>, index) => {
    e.preventDefault();
    e.stopPropagation();

    const updatedResult = document.getElementById(`${index}-dmn-result`);
    updatedResult?.classList.remove("kogito--editor__dmn-form-result__leaf-updated");
  }, []);

  const onOpenExecutionTab = useCallback(() => {
    if (props.notificationsPanel) {
      openExecutionTab?.();
    }
  }, [props.notificationsPanel, openExecutionTab]);

  const resultStatus = useCallback(
    (evaluationStatus: DmnEvaluationStatus) => {
      switch (evaluationStatus) {
        case DmnEvaluationStatus.SUCCEEDED:
          return (
            <>
              <div className={"kie-tools__dmn-form-result__evaluation"}>
                <CheckCircleIcon />
                {props.notificationsPanel ? (
                  <a onClick={onOpenExecutionTab} className={"kogito--editor__dmn-form-result__evaluation-link"}>
                    {i18n.result.evaluation.success}
                  </a>
                ) : (
                  <p className={"kogito--editor__dmn-form-result__evaluation-link"}>{i18n.result.evaluation.success}</p>
                )}
              </div>
            </>
          );
        case DmnEvaluationStatus.SKIPPED:
          return (
            <>
              <div className={"kie-tools__dmn-form-result__evaluation"}>
                <InfoCircleIcon />
                {props.notificationsPanel ? (
                  <a onClick={onOpenExecutionTab} className={"kogito--editor__dmn-form-result__evaluation-link"}>
                    {i18n.result.evaluation.skipped}
                  </a>
                ) : (
                  <p className={"kogito--editor__dmn-form-result__evaluation-link"}>{i18n.result.evaluation.skipped}</p>
                )}
              </div>
            </>
          );
        case DmnEvaluationStatus.FAILED:
          return (
            <>
              <div className={"kie-tools__dmn-form-result__evaluation"}>
                <ExclamationCircleIcon />
                {props.notificationsPanel ? (
                  <a onClick={onOpenExecutionTab} className={"kogito--editor__dmn-form-result__evaluation-link"}>
                    {i18n.result.evaluation.failed}
                  </a>
                ) : (
                  <p className={"kogito--editor__dmn-form-result__evaluation-link"}>{i18n.result.evaluation.failed}</p>
                )}
              </div>
            </>
          );
      }
    },
    [
      i18n.result.evaluation.failed,
      i18n.result.evaluation.skipped,
      i18n.result.evaluation.success,
      onOpenExecutionTab,
      props.notificationsPanel,
    ]
  );

  const result = useCallback(
    (dmnFormResult: DmnEvaluationResult, parentKey?: string) => {
      switch (typeof dmnFormResult) {
        case "boolean":
          return dmnFormResult ? <i>true</i> : <i>false</i>;
        case "number":
          return dmnFormResult;
        case "string":
          if (dmnFormResult.match(DATE_REGEX)) {
            const current = new Date(dmnFormResult);
            return (
              <>
                <Tooltip
                  key={`date-tooltip-${dmnFormResult}`}
                  content={
                    <I18nWrapped components={{ date: current.toString() }}>{i18n.result.dateTooltip}</I18nWrapped>
                  }
                >
                  <div className={"kogito--editor__dmn-form-result__results-date"}>
                    <p className={"kogito--editor__dmn-form-result__results-date"}>{dmnFormResult}</p>
                    <OutlinedQuestionCircleIcon />
                  </div>
                </Tooltip>
              </>
            );
          }
          return dmnFormResult;
        case "object":
          if (!dmnFormResult) {
            return <i>(null)</i>;
          }

          if (Array.isArray(dmnFormResult)) {
            if (dmnFormResult.length === 0) {
              return (
                <>
                  {parentKey && <DescriptionListTerm>{parentKey}</DescriptionListTerm>}
                  <i>(null)</i>
                </>
              );
            }
            return (
              <DescriptionList>
                <DescriptionListGroup
                  style={{
                    boxShadow: "0 0px 3px rgba(3, 3, 3, 0.15)",
                    padding: "10px",
                  }}
                >
                  {dmnFormResult.map((dmnResult, index) => (
                    <React.Fragment key={`array-result-${index}`}>
                      <DescriptionListTerm>{parentKey ? `${parentKey}-${index}` : index}</DescriptionListTerm>
                      <DescriptionListDescription>{result(dmnResult)}</DescriptionListDescription>
                    </React.Fragment>
                  ))}
                </DescriptionListGroup>
              </DescriptionList>
            );
          }
          return (
            <DescriptionList>
              <DescriptionListGroup
                style={{
                  boxShadow: "0 0px 3px rgba(3, 3, 3, 0.15)",
                  padding: "10px",
                }}
              >
                {Object.entries(dmnFormResult).map(([key, value]: [string, object | string]) => (
                  <React.Fragment key={`object-result-${key}-${value}`}>
                    {value === null && (
                      <DescriptionListTerm>{parentKey ? `${parentKey}-${key}` : key}</DescriptionListTerm>
                    )}
                    {value !== null && typeof value !== "object" && (
                      <DescriptionListTerm>{parentKey ? `${parentKey}-${key}` : key}</DescriptionListTerm>
                    )}
                    <DescriptionListDescription>
                      {result(value, parentKey ? `${parentKey}-${key}` : key)}
                    </DescriptionListDescription>
                  </React.Fragment>
                ))}
              </DescriptionListGroup>
            </DescriptionList>
          );

        default:
          return <i>(null)</i>;
      }
    },
    [i18n]
  );

  const resultsToRender = useMemo(
    () =>
      props.results?.map((dmnFormResult, index) => (
        <div key={`${index}-dmn-result`} className={"kogito--editor__dmn-form-result__results"}>
          <Card
            id={`${index}-dmn-result`}
            isFlat={true}
            className={"kogito--editor__dmn-form-result__results-card"}
            onAnimationEnd={(e) => onAnimationEnd(e, index)}
          >
            <CardTitle>
              <Title headingLevel={"h2"}>{dmnFormResult.decisionName}</Title>
            </CardTitle>
            <CardBody isFilled={true}>{result(dmnFormResult.result)}</CardBody>
            <CardFooter>{resultStatus(dmnFormResult.evaluationStatus)}</CardFooter>
          </Card>
        </div>
      )),
    [onAnimationEnd, props.results, result, resultStatus]
  );

  const formResultErrorMessage = useMemo(
    () => (
      <div>
        <EmptyState>
          <EmptyStateIcon icon={ExclamationTriangleIcon} />
          <TextContent>
            <Text component={"h2"}>{i18n.result.error.title}</Text>
          </TextContent>
          <EmptyStateBody>
            <TextContent>{i18n.result.error.explanation}</TextContent>
            <br />
            <TextContent>
              <I18nWrapped
                components={{
                  jira: (
                    <a href={ISSUES_URL} target={"_blank"} rel={"noopener noreferrer"}>
                      {ISSUES_URL}
                    </a>
                  ),
                }}
              >
                {i18n.result.error.message}
              </I18nWrapped>
            </TextContent>
          </EmptyStateBody>
        </EmptyState>
      </div>
    ),
    [i18n]
  );

  useEffect(() => {
    if (resultsToRender && resultsToRender.length > 0) {
      setFormResultStatus(FormDmnOutputsStatus.VALID);
    } else if (formResultError) {
      setFormResultStatus(FormDmnOutputsStatus.ERROR);
    } else {
      setFormResultStatus(FormDmnOutputsStatus.EMPTY);
    }
  }, [resultsToRender, formResultError]);

  // Resets the ErrorBoundary everytime the FormSchema is updated
  useEffect(() => {
    errorBoundaryRef.current?.reset();
  }, [props.results]);

  return (
    <>
      {formResultStatus === FormDmnOutputsStatus.EMPTY && (
        <EmptyState>
          <EmptyStateIcon icon={InfoCircleIcon} />
          <TextContent>
            <Text component={"h2"}>{i18n.result.withoutResponse.title}</Text>
          </TextContent>
          <EmptyStateBody>
            <TextContent>
              <Text>{i18n.result.withoutResponse.explanation}</Text>
            </TextContent>
          </EmptyStateBody>
        </EmptyState>
      )}
      {formResultStatus === FormDmnOutputsStatus.ERROR && formResultErrorMessage}
      {formResultStatus === FormDmnOutputsStatus.VALID && (
        <ErrorBoundary ref={errorBoundaryRef} setHasError={setFormResultError} error={formResultErrorMessage}>
          <div data-testid={"dmn-form-result"}>{resultsToRender}</div>
        </ErrorBoundary>
      )}
    </>
  );
}
