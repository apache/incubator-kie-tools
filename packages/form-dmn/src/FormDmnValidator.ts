/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { Validator } from "@kie-tools/form/dist/Validator";
import { FormDmnI18n } from "./i18n";
import { FormDmnJsonSchemaBridge } from "./uniforms";
import {
  DAYS_AND_TIME_DURATION_FORMAT,
  X_DMN_ALLOWED_VALUES_KEYWORD,
  YEARS_AND_MONTHS_DURATION_FORMAT,
} from "@kie-tools/dmn-runner/dist/constants";
import { ExtendedServicesDmnJsonSchema } from "@kie-tools/extended-services-api";
import { DmnRunnerAjv } from "@kie-tools/dmn-runner/dist/ajv";
import { SCHEMA_DRAFT4 } from "@kie-tools/dmn-runner/dist/constants";

export class FormDmnValidator extends Validator {
  private dmnRunnerAjv = new DmnRunnerAjv();

  constructor(i18n: FormDmnI18n) {
    super(i18n);
  }

  // Override to add period validation
  public createValidator(jsonSchema: any) {
    const validator = this.dmnRunnerAjv.getAjv().compile(jsonSchema);

    return (model: any) => {
      // AJV doesn't handle dates objects. This transformation converts Dates to their UTC format.
      validator(JSON.parse(JSON.stringify(model)));

      if (!validator.errors?.length) {
        return null;
      }

      return {
        details: validator.errors?.map((error: any) => {
          if (error.keyword === "format") {
            if ((error.params as any).format === DAYS_AND_TIME_DURATION_FORMAT) {
              return { ...error, message: (this.i18n as FormDmnI18n).validation.daysAndTimeError };
            }
            if ((error.params as any).format === YEARS_AND_MONTHS_DURATION_FORMAT) {
              return { ...error, message: (this.i18n as FormDmnI18n).validation.yearsAndMonthsError };
            }
          }
          if (error.keyword === X_DMN_ALLOWED_VALUES_KEYWORD) {
            return { ...error, message: (this.i18n as FormDmnI18n).validation.xDmnAllowedValues };
          }
          return error;
        }),
      };
    };
  }

  public getBridge(formSchema: ExtendedServicesDmnJsonSchema): FormDmnJsonSchemaBridge {
    const formDraft4 = { ...formSchema, $schema: SCHEMA_DRAFT4 };
    const validator = this.createValidator(formDraft4);
    return new FormDmnJsonSchemaBridge(formDraft4, validator, this.i18n as FormDmnI18n);
  }
}
