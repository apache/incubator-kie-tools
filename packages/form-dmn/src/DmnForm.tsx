/*
 * Copyright 2022 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from "react";
import { useMemo } from "react";
import { DmnValidator } from "./DmnValidator";
import { dmnFormI18n } from "./i18n";
import { FormComponent, FormComponentProps, FormProps } from "@kie-tools/form";
import { DmnAutoFieldProvider } from "./uniforms/DmnAutoFieldProvider";

export type InputRow = Record<string, string>;

type DmnDecisionNodes = "InputSet" | string;

export interface DmnSchema {
  definitions?: Record<
    DmnDecisionNodes,
    {
      type: string;
      properties: Record<string, DmnDeepProperty>;
    }
  >;
}

export interface DmnFormData {
  definitions: DmnFormDefinitions;
}

type DmnFormDefinitions = Record<
  DmnDecisionNodes,
  {
    required?: string[];
    properties: object;
    type: string;
    placeholder?: string;
    title?: string;
    format?: string;
    items: object[] & { properties: object };
    "x-dmn-type"?: string;
  }
>;

interface DmnDeepProperty {
  $ref?: string;
  type?: string;
  placeholder?: string;
  title?: string;
  format?: string;
  "x-dmn-type"?: string;
  properties?: DmnDeepProperty;
}

export function DmnForm(props: FormProps<InputRow, DmnSchema>) {
  const i18n = useMemo(() => {
    dmnFormI18n.setLocale(props.locale ?? navigator.language);
    return dmnFormI18n.getCurrent();
  }, [props.locale]);
  const dmnValidator = useMemo(() => new DmnValidator(i18n), [i18n]);

  return (
    <FormComponent
      {...props}
      i18n={i18n}
      validator={dmnValidator}
      removeRequired={true}
      entryPath={"definitions.InputSet"}
      propertiesEntryPath={"definitions.InputSet.properties"}
    >
      <DmnAutoFieldProvider />
    </FormComponent>
  );
}
