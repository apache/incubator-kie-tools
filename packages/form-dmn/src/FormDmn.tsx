/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import * as React from "react";
import { useMemo } from "react";
import { FormDmnValidator } from "./FormDmnValidator";
import { formDmnI18n } from "./i18n";
import { FormComponent, FormProps } from "@kie-tools/form/dist/FormComponent";
import { DmnAutoFieldProvider } from "@kie-tools/dmn-runner/dist/uniforms";
import { ExtendedServicesDmnJsonSchema } from "@kie-tools/extended-services-api";
import { formDmnRunnerAutoFieldValue } from "./uniforms/FormDmnRunnerAutoFieldValue";

export type InputRow = Record<string, any>;

export function FormDmn(props: FormProps<InputRow, ExtendedServicesDmnJsonSchema>) {
  const i18n = useMemo(() => {
    formDmnI18n.setLocale(props.locale ?? navigator.language);
    return formDmnI18n.getCurrent();
  }, [props.locale]);
  const dmnValidator = useMemo(() => new FormDmnValidator(i18n), [i18n]);

  return (
    <FormComponent
      {...props}
      i18n={i18n}
      validator={dmnValidator}
      removeRequired={true}
      entryPath={"definitions.InputSet"}
      propertiesEntryPath={"definitions.InputSet.properties"}
    >
      <DmnAutoFieldProvider value={formDmnRunnerAutoFieldValue} />
    </FormComponent>
  );
}
