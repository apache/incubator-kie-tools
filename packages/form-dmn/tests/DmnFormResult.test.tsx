/*
 * Copyright 2022 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from "react";
import { render } from "@testing-library/react";
import { DecisionResult, DmnFormResult, DmnFormResultProps, EvaluationStatus } from "../src";

const props: DmnFormResultProps = {
  results: [],
  differences: [{}],
  locale: "en",
  notificationsPanel: true,
  openExecutionTab: () => {},
};

describe("DmnFormResult tests", () => {
  it("should render the DMNFormResult with the empty state", async () => {
    const { getByText } = render(<DmnFormResult {...props} />);

    expect(getByText("No response")).toMatchSnapshot();
  });

  it("should render the DmnFormResult with one result", async () => {
    const results: DecisionResult[] = [
      {
        decisionId: "_9BD7BB23-0B23-488F-8DED-F5462CF89E0B",
        decisionName: "Decision-1",
        result: null,
        messages: [],
        evaluationStatus: EvaluationStatus.FAILED,
      },
    ];

    const { getByText } = render(<DmnFormResult {...props} results={results} />);

    expect(getByText("Decision-1")).toMatchSnapshot();
  });

  it("should render the DmnFormResult with more then one result", async () => {
    const results: DecisionResult[] = [
      {
        decisionId: "_9BD7BB23-0B23-488F-8DED-F5462CF89E0B",
        decisionName: "Decision-1",
        result: null,
        messages: [],
        evaluationStatus: EvaluationStatus.FAILED,
      },
      {
        decisionId: "_9BD7BB23-0B23-488F-8DED-F5462CF89E0B",
        decisionName: "Decision-2",
        result: null,
        messages: [],
        evaluationStatus: EvaluationStatus.SUCCEEDED,
      },
      {
        decisionId: "_9BD7BB23-0B23-488F-8DED-F5462CF89E0B",
        decisionName: "Decision-3",
        result: null,
        messages: [],
        evaluationStatus: EvaluationStatus.SKIPPED,
      },
    ];

    const { getByText } = render(<DmnFormResult {...props} results={results} />);

    expect(getByText("Decision-1")).toMatchSnapshot();
    expect(getByText("Decision-2")).toMatchSnapshot();
    expect(getByText("Decision-3")).toMatchSnapshot();
  });

  it("should render an anchor tag", async () => {
    const results: DecisionResult[] = [
      {
        decisionId: "_9BD7BB23-0B23-488F-8DED-F5462CF89E0B",
        decisionName: "Decision-1",
        result: null,
        messages: [],
        evaluationStatus: EvaluationStatus.FAILED,
      },
    ];

    const { getByText } = render(<DmnFormResult {...props} results={results} />);

    expect(getByText("Evaluation failed").tagName).toEqual("A");
    expect(getByText("Evaluation failed")).toMatchSnapshot();
  });

  it("should render an paragraph tag", async () => {
    const results: DecisionResult[] = [
      {
        decisionId: "_9BD7BB23-0B23-488F-8DED-F5462CF89E0B",
        decisionName: "Decision-1",
        result: null,
        messages: [],
        evaluationStatus: EvaluationStatus.FAILED,
      },
    ];

    const { getByText } = render(<DmnFormResult {...props} notificationsPanel={false} results={results} />);

    expect(getByText("Evaluation failed").tagName).toEqual("P");
    expect(getByText("Evaluation failed")).toMatchSnapshot();
  });
});
