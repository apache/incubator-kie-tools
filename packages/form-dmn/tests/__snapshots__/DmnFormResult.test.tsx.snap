// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DmnFormResult tests should render an anchor tag 1`] = `
<a
  class="kogito--editor__dmn-form-result__evaluation-link"
>
  Evaluation failed
</a>
`;

exports[`DmnFormResult tests should render an paragraph tag 1`] = `
<p
  class="kogito--editor__dmn-form-result__evaluation-link"
>
  Evaluation failed
</p>
`;

exports[`DmnFormResult tests should render the DMNFormResult with the empty state 1`] = `
<h2
  class=""
  data-ouia-component-id="OUIA-Generated-Text-1"
  data-ouia-component-type="PF4/Text"
  data-ouia-safe="true"
  data-pf-content="true"
>
  No response
</h2>
`;

exports[`DmnFormResult tests should render the DmnFormResult with more then one result 1`] = `
<h2
  class="pf-c-title pf-m-xl"
  data-ouia-component-id="OUIA-Generated-Title-2"
  data-ouia-component-type="PF4/Title"
  data-ouia-safe="true"
>
  Decision-1
</h2>
`;

exports[`DmnFormResult tests should render the DmnFormResult with more then one result 2`] = `
<h2
  class="pf-c-title pf-m-xl"
  data-ouia-component-id="OUIA-Generated-Title-3"
  data-ouia-component-type="PF4/Title"
  data-ouia-safe="true"
>
  Decision-2
</h2>
`;

exports[`DmnFormResult tests should render the DmnFormResult with more then one result 3`] = `
<h2
  class="pf-c-title pf-m-xl"
  data-ouia-component-id="OUIA-Generated-Title-4"
  data-ouia-component-type="PF4/Title"
  data-ouia-safe="true"
>
  Decision-3
</h2>
`;

exports[`DmnFormResult tests should render the DmnFormResult with one result 1`] = `
<h2
  class="pf-c-title pf-m-xl"
  data-ouia-component-id="OUIA-Generated-Title-1"
  data-ouia-component-type="PF4/Title"
  data-ouia-safe="true"
>
  Decision-1
</h2>
`;
